### PR TITLE
Filesystem adapter for GoogleDrive

### PR DESCRIPTION
**Pull request by Arno2005**

### CODE REVIEW

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveApp.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            'state' => csrf_token(),
+            'scope' => self::AUTH_SCOPE,
+            'redirect_uri' => $redirectUrl,
+            'access_type' => $accessType,
+        ];
+
+        return self::AUTH_URL . '?' . http_build_query($params, '', '&');
+    }
+
+    public function getAndSaveAccessToken($code, $redirectUrl = '', $byRefresh = false): object
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveApp.php:90]: Name this function to `fetchTokens()` to correspond to DroboxApp naming convention

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveApp.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            ->start();
+
+
+        $errors = $this->httpClient->getErrors();
+        $responseBody = $this->httpClient->getResponseBody();
+
+        if ($errors) {
+            $code = $errors['code'];
+
+            if ($code == 401) {
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveApp.php:130]: Avoid using magic numbers and try to use appropriate constants that available

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+class GoogleDriveFileSystemAdapter implements FilesystemAdapterInterface
+{
+
+    /**
+     * @var GoogleDriveFileSystemAdapter|null
+     */
+    private static $instance = null;
+
+    private $googleDriveApp;
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:16]: Add the dock-block for the property

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    public static function getInstance(GoogleDriveApp $googleDriveApp): GoogleDriveFileSystemAdapter
+    {
+        if (self::$instance === null) {
+            self::$instance = new self($googleDriveApp);
+        }
+
+        return self::$instance;
+    }
+
+    public function makeDirectory($dirname, $parentId = null): bool
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:41]: Add type hints for method paramteres

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            return (string)$this->googleDriveApp->getFileInfo($filename, true);
+        } catch (Exception $e) {
+            return false;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function put(string $filename, string $content, string $parentId = null)
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:80]: Add question mark for `$parentId`

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        try {
+            if($this->isFile($filename)){
+                return $this->googleDriveApp->rpcRequest(GoogleDriveApp::FILE_MEDIA_URL . '/' . $filename . '?uploadType=media', $content, 'PATCH', 'application/octet-stream');
+            }else{
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:85]: Is this really necessary to do this, doesn't google create new file (if not exists) or overwrite the file (if exists) ?

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function isFile(string $filename): bool
+    {
+        try {
+            $meta = (array)$this->googleDriveApp->getFileInfo($filename);
+
+            return $meta['kind'] === "drive#file" && $meta['mimeType'] != GoogleDriveApp::FOLDER_MIMETYPE;
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:201]: use as constants `drive#file`

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function isDirectory(string $dirname): bool
+    {
+        try {
+            $meta = (array)$this->googleDriveApp->getFileInfo($dirname);
+
+            return $meta['kind'] === "drive#file" && $meta['mimeType'] === GoogleDriveApp::FOLDER_MIMETYPE;
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:215]: What is the difference between file and folder?

Comment (armanist) [reply src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:215]: Looks like `isDirectory()` and `isFile()` cehcking the same thing

File: src/Libraries/Storage/Adapters/Local/LocalFileSystemAdapter.php [hunk 9]
```diff
@@ -43,9 +43,11 @@ public static function getInstance(): ?LocalFileSystemAdapter
     }
 
     /**
+     * @param string $dirname
+     * @param string|null $parentId
      * @inheritDoc
      */
-    public function makeDirectory(string $dirname): bool
+    public function makeDirectory(string $dirname, string $parentId = null): bool
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/Local/LocalFileSystemAdapter.php:50]: `?` question mark

File: src/Libraries/Storage/Adapters/Local/LocalFileSystemAdapter.php [hunk 10]
```diff
@@ -89,9 +91,12 @@ public function getJson(string $filename)
     }
 
     /**
+     * @param string $filename
+     * @param string $content
+     * @param string|null $parentId
      * @inheritDoc
      */
-    public function put(string $filename, string $content)
+    public function put(string $filename, string $content, string $parentId = null)
```
Comment (armanist) [inline src/Libraries/Storage/Adapters/Local/LocalFileSystemAdapter.php:99]: ? question mark

File: src/Libraries/Storage/FileSystem.php [hunk 11]
```diff
@@ -21,10 +21,10 @@
 /**
  * Class FileSystem
  * @package Quantum\Libraries\Storage
- * @method bool makeDirectory(string $dirname)
+ * @method bool makeDirectory(string $dirname, string $parentId = null)
```
Comment (armanist) [inline src/Libraries/Storage/FileSystem.php:24]: @method bool|null makeDirectory(string $dirname, ?string $parentId = null)

**Review by armanist — CHANGES_REQUESTED** [review]

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php (hunk 6 above)
Comment (Arno2005) [reply src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:85]: Google Drive itself doesn't support that, when overriding a file, you have to specify the file's id, whilst creating one requires a file name.

File: src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php (hunk 8 above)
Comment (Arno2005) [reply src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php:215]: Basically google drive folder is a file with its own mimetype (see the constant). isDirectory() checks if the file has that mimetype, on the other hand isFile() checks to see if the file doesn't have that mimetype.

**Review by armanist — APPROVED** [review]

File: tests/unit/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapterTest.php [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        $this->fs->append($this->filename, $moreContent);
+
+        $this->assertEquals($this->content . $moreContent, $this->fs->get($this->filename));
+    }
+
+    public function testGoogleDriveFileRename()
+    {
+        $this->fs->put($this->filename, $this->content);
+
+        $newFilename = 'new_name.txt';
```
Comment (Vahram1995) [inline tests/unit/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapterTest.php:99]: Replace to class property "private $newFilename = 'new_name.txt';"

**Review by armanist — APPROVED** [review]

**Pull request merged**